tls/s2n_handshake_io.c

```diff
@@ -1057,7 +1066,30 @@ static int s2n_handshake_read_io(struct s2n_connection *conn)
         record_type = TLS_HANDSHAKE;
         POSIX_GUARD_RESULT(s2n_quic_read_handshake_message(conn, &message_type));
     } else {
-        POSIX_GUARD(s2n_read_full_record(conn, &record_type, &isSSLv2));
+        int r = s2n_read_full_record(conn, &record_type, &isSSLv2);
+
+        /**
+         *= https://tools.ietf.org/rfc/rfc8446#section-4.2.10
+         *# If the client attempts a 0-RTT handshake but the server
+         *# rejects it, the server will generally not have the 0-RTT record
+         *# protection keys and must instead use trial decryption (either with
+         *# the 1-RTT handshake keys or by looking for a cleartext ClientHello in
+         *# the case of a HelloRetryRequest) to find the first non-0-RTT message.
+         *#
+         *# If the server chooses to accept the "early_data" extension, then it
+         *# MUST comply with the same error-handling requirements specified for
+         *# all records when processing early data records.  Specifically, if the
+         *# server fails to decrypt a 0-RTT record following an accepted
+         *# "early_data" extension, it MUST terminate the connection with a
+         *# "bad_record_mac" alert as per Section 5.2.
+         */
+        if ((r < 0) && (s2n_errno == S2N_ERR_DECRYPT) /* Decryption Error */
+                && (conn->mode == S2N_SERVER) /* On the server */
+                && (conn->early_data_state == S2N_EARLY_DATA_REJECTED) /* When early data was rejected */) {
+            POSIX_GUARD_RESULT(s2n_wipe_record(conn));
```

Comment:
Q: In the case that s2n_errno is not set to `S2N_ERR_DECRYPT` i.e decryption passes ( using trial decryption to find the first non-0-RTT message as described in the first paragraph) and  early data is rejected, we should succeed, right? Could we add a test case for this use-case.

Reply:
Spoke offline: Currently, writing this test requires large amount of setup code that far exceeds the value of the unit test as it will be covered in future self-talk/functional tests. This use-case will be tested in self-talk/functional tests when we add them for 0-RTT.

Reply:
If you're suggesting we test the "negotiate a normal handshake after rejecting early data", I believe that's out of scope here. It's one of the standard early data success cases and will be covered by functional tests, but testing just that case here would take a LOT of test setup (valid encryption keys, valid record, valid connection that can successfully trigger the handler for that record) and not provide any additional value. We should focus on failure cases here.